### ISSUE TITLE
Inconsistent Data Hashing from CLI & Serialization Lib?

### ISSUE THREAD

**Issue by stephenhuh** [Emurgo/cardano-serialization-lib#247]
Hey there!

having a similar issue to some other issues open, but i'm serializing an **empty** datum with just the constructor set but yet the hashes still don't line up. See examples below, almost no data is being passed through besides the `tag`. 

to be specific:

serialization lib:
```javascript
  const datum = Loader.Cardano.PlutusData.new_constr_plutus_data(
    Loader.Cardano.ConstrPlutusData.new(
      Loader.Cardano.Int.new_i32(0),
      Loader.Cardano.PlutusList.new()
    )
  );
```

then on the cli
```bash
cardano-cli transaction hash-script-data \
  --script-data-file scripts/datum.json 
```

where datum.json is
```json
{"constructor":0,"fields":[]}
```

i may be overlooking something but do let me know! otherwise i can't seem to figure out where the difference is.

the hashes respectively are:
`f7f2f57c58b5e4872201ab678928b0d63935e82d022d385e1bad5bfe347e89d8`
`923918e403bf43c34b4ef6b48eb2ee04babed17320d8d1b9ff9ad086e86f44ec`

--- UPDATE: ---
for additional context i get the first hash by loading it into a `TransactionOutput` then calling `.data_hash().to_bytes()` then, hex encoding the byte output. the same output is achieved by just calling `hash_plutus_data()` then converting `.to_bytes()` then hex encoding.

**Comment by vsubhuman** [comment]
@rooooooooob , plz check when possible

**Comment by vsubhuman** [comment]
Might be related to https://github.com/Emurgo/cardano-serialization-lib/pull/217, potentially, so need to be rechecked in the next version

**Comment by m-a-c-k** [comment]
crucial for SC development ^

**Comment by JonathanPynes** [comment]
Hmmmmm

**Comment by rooooooooob** [comment]
I'm going to try and look into this more. There's multiple problems and the fact that cardano-node doesn't (will change next hardfork at least in langdepview see input-output-hk/cardano-ledger-specs#2512) follow the specs in some places makes it a bit tricky. From what it looks like in general their serialization is different as well e.g. for plutus data #228 

On top of that there's the issue with canonical ordering which we've got an open PR (#246) here in one area but I'll need to investigate where else might need canonical ordering, and whether or not cardano-node follows the specs there.

Is there any way you can provide me with the datum itself (not the hash) from cardano-node? It can be hex'd cbor bytes or whatever. Although that might not be available without modifying the node if you're passing in just JSON. It's weird since I thought they were using definite encoding for empty lists/maps and only indefinite for non-empty ones, whereas we do definite for all.

**Comment by jfischoff** [comment]
@rooooooooob isn't this the datum?

```json
{"constructor":0,"fields":[]}
```

**Comment by rooooooooob** [comment]
@jfischoff Yeah but I meant how they encode that, since that's some cardano-node specific JSON format I think. I'll need to investigate their code then to see how they turn that into CBOR to serialized into a binary string, since that is what the hash is computed on.

**Comment by jfischoff** [comment]
Looking at some other examples I can see where this stems from. When encoding the plutus data the cli uses construct tag 121 and then a list. The serialization lib uses tag 102 and then a list with a uint and another list. 

You can see both alternatives are considered valid constructors: https://github.com/input-output-hk/cardano-ledger-specs/blob/master/eras/alonzo/test-suite/cddl-files/alonzo.cddl#L274

**Comment by rooooooooob** [comment]
That's weird though. The 102 tag is for tags that don't fall into the 121-127 and 1280-1400 range, which 0 does not. So it should be 102 in the cardano-node too, not 121. Either that or I don't understand what the `"constructor"` field in that JSON is supposed to mean.

**Comment by jfischoff** [comment]
I agree. What I am seeing from the CLI, and you can try on your end, is it is treating 121 as index 0 and 122 as index 1.

**Comment by rooooooooob** [comment]
@jfischoff was indeed correct. The problem was we based the code of this off of the .cddl binary spec, but if you look at https://github.com/input-output-hk/plutus/blob/1f31e640e8a258185db01fa899da63f9018c0e85/plutus-core/plutus-core/src/PlutusCore/Data.hs#L61 you'll see that there's actually a difference between the conceptual constructor alternative marker and the CBOR tag it is encoded to when it's a compact form one.

>The scheme is:
>- Alternatives 0-6 -> tags 121-127, followed by the arguments in a list
>- Alternatives 7-127 -> tags 1280-1400, followed by the arguments in a list

Since we worked off of the binary spec we exposed the CBOR tag in the compact form, but since it seems those aren't a 1-to-1 mapping in the compact tag form I think it makes more sense to instead just expose the and leave the tag as an implementation detail in (de)serialization code.

@stephenhuh I have a fix for this in #250
